devito/finite_differences/differentiable.py

```diff
@@ -0,0 +1,182 @@
+            assert len(args) == 1
+            expr = args[0]
+            new_obj = sympy.Expr.__new__(cls, expr)
+            # Initialization
+            new_obj.__init__(expr, **kwargs)
+            return new_obj
+        return sympy.Expr.__new__(cls, *args)
+
+    def __init__(self, expr, **kwargs):
+        # Recover the list of possible FD shortcuts
+        self._dtype = kwargs.get('dtype')
+        self._space_order = kwargs.get('space_order')
+        self._time_order = kwargs.get('time_order', 0)
+        self._indices = kwargs.get('indices', ())
+        self._staggered = kwargs.get('staggered')
+        self._grid = kwargs.get('grid')
+        # Generate FD shortcuts for expression or copy from input
+        self._fd = kwargs.get('fd', {})
+        # Save kwargs
+        self._expr = expr
+
+    @property
+    def space_order(self):
+        return self._space_order
+
+    @property
+    def time_order(self):
+        return self._time_order
+
+    @property
+    def staggered(self):
+        return self._staggered
+
+    @property
+    def indices(self):
+        return self._indices
+
+    @property
+    def fd(self):
+        return self._fd
+
+    @property
+    def grid(self):
+        return self._grid
+
+    @property
+    def dtype(self):
+        return self._dtype
+
+    def __getattr__(self, name):
+        if name == 'fd' or name == '_fd':
+            raise AttributeError()
+        if name in self.fd.keys():
+            return self.fd[name][0](self)
+        return self.__getattribute__(name)
+
+    def xreplace(self, rule):
+        out = getattr(self, '_expr', self)
+        if out in rule:
+            return rule[out]
+        elif rule:
+            args = []
+            for a in out.args:
+                try:
+                    args.append(a.xreplace(rule))
+                except AttributeError:
+                    args.append(a)
+            args = tuple(args)
+            if not _aresame(args, out.args):
+                return out.func(*args)
+        return out
+
+    def _merge_fd_properties(self, other):
+        merged = {}
+        merged["space_order"] = np.min([getattr(self, 'space_order', 100) or 100,
+                                        getattr(other, 'space_order', 100)])
+        merged["time_order"] = np.min([getattr(self, 'time_order', 100) or 100,
+                                       getattr(other, 'time_order', 100)])
+        merged["indices"] = tuple(set(self.indices + getattr(other, 'indices', ())))
+        merged["fd"] = dict(getattr(self, 'fd', {}), **getattr(other, 'fd', {}))
+        merged["staggered"] = self.staggered
+        merged["dtype"] = self.dtype
+        return merged
+
+    def __add__(self, other):
+        return Differentiable(sympy.Add(*[getattr(self, '_expr', self),
+                                          getattr(other, '_expr', other)]),
+                              **self._merge_fd_properties(other))
+
+    __iadd__ = __add__
```

Comment:
I'm not sure __iadd__ and __radd__ should be the same as __add__. Is it what happens in sympy typically? __iadd__ is += 

same for sub, mul etc

Reply:
Sympy expressions are immutable, and operation on it should return a new one.

Reply:
it appears that += actually has an effect:
```
In [1]: from sympy import *
...
In [5]: a=Symbol('a')
In [6]: b=Symbol('b')
In [7]: a + b
Out[7]: a + b
In [8]: a += b
In [9]: a
Out[9]: a + b
```
note that the content of ``a`` has actually changed

as for __radd__, __rmul__, __rdiv__ ... we should swap the operands. See for example here: https://github.com/sympy/sympy/blob/master/sympy/core/expr.py#L124
not sure, however, what are the practical implications of this in SymPy

Reply:
Hum odd, I can rewrite `iadd,imul,..` to just change `_expr` in place and keep same object, fine with me.

for the `r` ones, additions are commutative so doesn't really matter but can change it.